src/sentry/release_health/duplex.py

```diff
@@ -0,0 +1,763 @@
+        ]
+
+    if schema is None:
+        iterate_all = True
+        generic_item_schema = None
+        schema = {}
+    else:
+        iterate_all = "*" in schema
+        generic_item_schema = schema.get("*")
+
+    ret_val = []
+
+    if iterate_all:
+        if len(sessions) != len(metrics):
+            return [
+                f"different number of keys in dictionaries sessions={len(sessions)}, metrics={len(metrics)}"
+            ]
+        for key, val in sessions.items():
+            child_path = f"{path}[{key}]"
+            child_schema = schema.get(key, generic_item_schema)
+            ret_val += compare_results(val, metrics.get(key), rollup, child_path, child_schema)
+    else:
+        for key, child_schema in schema.items():
+            child_path = f"{path}[{key}]"
+            ret_val += compare_results(
+                sessions.get(key), metrics.get(key), rollup, child_path, child_schema
+            )
+    return ret_val
+
+
+def compare_results(
+    sessions, metrics, rollup: int, path: Optional[str] = None, schema: Optional[Schema] = None
+) -> List[str]:
+    if path is None:
+        path = ""
+
+    if schema is not None:
+        discriminator = schema
+    else:
+        discriminator = sessions
+
+    if discriminator is None:
+        if metrics is None:
+            return []
+        else:
+            return [f"unmatched field at path {path}, sessions=None, metrics={metrics}"]
+
+    if type(discriminator) in {str, float, int, datetime, ComparatorType}:
+        err = compare_scalars(sessions, metrics, rollup, path, schema)
+        if err is not None:
+            return [err]
+        else:
+            return []
+    elif type(discriminator) == tuple:
+        return compare_tuples(sessions, metrics, rollup, path, schema)
+    elif type(discriminator) == list:
+        return compare_arrays(sessions, metrics, rollup, path, schema)
+    elif type(discriminator) == set:
+        return compare_sets(sessions, metrics, path)
+    elif type(discriminator) == dict:
+        return compare_dicts(sessions, metrics, rollup, path, schema)
+    else:
+        return [f"invalid schema type={type(schema)} at path:'{path}'"]
+
+
+class DuplexReleaseHealthBackend(ReleaseHealthBackend):
+    DEFAULT_ROLLUP = 60 * 60  # 1h
+
+    def __init__(
+        self,
+        metrics_start: datetime,
+    ):
+        self.sessions = SessionsReleaseHealthBackend()
+        self.metrics = MetricsReleaseHealthBackend()
+        self.metrics_start = metrics_start
+
+    def log_exception(self, ex, result=None):
+        pass  # TODO
+
+    def log_errors(self, errors, sessions, metrics):
+        pass
+
+    def _dispatch_call(
+        self,
+        fn_name: str,
+        should_compare: Union[bool, Callable[[Any], bool]],
+        rollup: int,
+        schema: Optional[Schema],
+        *args,
+    ):
+        sessions_fn = getattr(self.sessions, fn_name)
+        ret_val = sessions_fn(*args)
```

Comment:
Should we check a feature flag at this point, so we can turn on the double calling only for certain projects?

Reply:
Yes.. not fully finished